File: python/nve/hub.py
"""
NVE Hub — download and resolve HuggingFace models.

Provides a Python interface for downloading models from HuggingFace Hub
and resolving model IDs to local paths.
"""

from __future__ import annotations

import json
import os
import subprocess
import shutil
from pathlib import Path
from typing import Optional


def resolve_model(
    model_id_or_path: str,
    cache_dir: Optional[str | Path] = None,
) -> Path:
    """
    Resolve a model ID or local path to a directory containing model files.

    Args:
        model_id_or_path: Either a local directory path or a HuggingFace model ID
                          (e.g. "meta-llama/Llama-3.2-1B", "microsoft/phi-2").
        cache_dir: Optional cache directory. Defaults to ~/.cache/nve/models.

    Returns:
        Path to the model directory containing config.json, tokenizer.json, and safetensors.
    """
    path = Path(model_id_or_path)

    # If it's already a local directory with config.json, use it.
    if path.is_dir() and (path / "config.json").exists():
        return path

    # Parse as a HuggingFace model ID.
    cache = Path(cache_dir) if cache_dir else _default_cache_dir()
    model_cache = cache / "models" / model_id_or_path.replace("/", "--")

    # Check cache.
    if model_cache.is_dir() and (model_cache / "config.json").exists():
        return model_cache

    # Download.
    print(f"Downloading model '{model_id_or_path}' to {model_cache}")
    _download_model(model_id_or_path, model_cache)
    return model_cache


def detect_architecture(model_dir: str | Path) -> dict:
    """
    Detect the architecture of a model from its config.json.

    Returns a dict with architecture info including:
    - arch: Architecture name (e.g. "llama", "gpt2", "mistral")
    - hidden_size, num_layers, num_heads, etc.
    """
    model_dir = Path(model_dir)
    config_path = model_dir / "config.json"
    if not config_path.exists():
        raise FileNotFoundError(f"No config.json found in {model_dir}")

    with open(config_path) as f:
        config = json.load(f)

    # Detect architecture.
    arch = _detect_arch(config)

    return {
        "arch": arch,
        "model_type": config.get("model_type", "unknown"),
        "hidden_size": config.get("hidden_size", config.get("n_embd")),
        "num_layers": config.get("num_hidden_layers", config.get("n_layer")),
        "num_heads": config.get("num_attention_heads", config.get("n_head")),
        "num_kv_heads": config.get("num_key_value_heads"),
        "vocab_size": config.get("vocab_size"),
        "intermediate_size": config.get("intermediate_size", config.get("n_inner")),
        "max_position_embeddings": config.get(
            "max_position_embeddings", config.get("n_positions")
        ),
    }


def _detect_arch(config: dict) -> str:
    """Detect architecture from config dict."""
    architectures = config.get("architectures", [])
    model_type = config.get("model_type", "")

    for a in architectures:
        a_lower = a.lower()
        if "llama" in a_lower:
            return "llama"
        if "mistral" in a_lower:
            return "mistral"
        if "qwen2" in a_lower:
            return "qwen2"
        if "phi3" in a_lower or "phiforcausal" in a_lower:
            return "phi3"
        if "gemma2" in a_lower:
            return "gemma2"
        if "gemma" in a_lower:
            return "gemma"
        if "gptneox" in a_lower:
            return "gpt_neox"
        if "gpt2" in a_lower:
            return "gpt2"
        if "falcon" in a_lower or "rwforcausal" in a_lower:
            return "falcon"
        if "stablelm" in a_lower:
            return "stablelm"
        if "starcoder2" in a_lower:
            return "starcoder2"
        if "internlm2" in a_lower:
            return "internlm2"
        if "olmo" in a_lower:
            return "olmo"
        if "deepseek" in a_lower:
            return "deepseek"

    # Fallback to model_type.
    return model_type or "unknown"


def _default_cache_dir() -> Path:
    hf_home = os.environ.get("HF_HOME")
    if hf_home:
        return Path(hf_home) / "nve"
    xdg = os.environ.get("XDG_CACHE_HOME")
    if xdg:
        return Path(xdg) / "nve"
    return Path.home() / ".cache" / "nve"


def _download_model(model_id: str, dest: Path):
    """Download model files from HuggingFace Hub."""
    dest.mkdir(parents=True, exist_ok=True)

    # Try huggingface-cli first.
    if shutil.which("huggingface-cli"):
        result = subprocess.run(
            [
                "huggingface-cli", "download", model_id,
                "--local-dir", str(dest),
                "--include", "config.json",
                "--include", "tokenizer.json",
                "--include", "tokenizer_config.json",
                "--include", "*.safetensors",
                "--include", "*.safetensors.index.json",
            ],
            capture_output=True,
            text=True,
        )
        if result.returncode == 0:
            return

    # Try huggingface_hub Python package.
    try:
        from huggingface_hub import snapshot_download
        snapshot_download(
            model_id,
            local_dir=str(dest),
            allow_patterns=[
                "config.json",
                "tokenizer.json",
                "tokenizer_config.json",
                "*.safetensors",
                "*.safetensors.index.json",
            ],
        )
        return
    except ImportError:
        pass

    # Fall back to curl.
    revision = "main"
    base_url = f"https://huggingface.co/{model_id}/resolve/{revision}"
    token = os.environ.get("HF_TOKEN", os.environ.get("HUGGING_FACE_HUB_TOKEN"))

    for filename in ["config.json", "tokenizer.json"]:
        _curl_download(f"{base_url}/{filename}", dest / filename, token)

    for filename in ["tokenizer_config.json", "model.safetensors", "model.safetensors.index.json"]:
        try:
            _curl_download(f"{base_url}/{filename}", dest / filename, token)
        except Exception:
            pass

    # Download shards if index exists.
    index_path = dest / "model.safetensors.index.json"
    if index_path.exists():
        with open(index_path) as f:
            index = json.load(f)
        shards = set(index.get("weight_map", {}).values())
        for shard in shards:
            shard_path = dest / shard
            if not shard_path.exists():
                _curl_download(f"{base_url}/{shard}", shard_path, token)


def _curl_download(url: str, dest: Path, token: Optional[str] = None):
    """Download a file using curl."""
    cmd = ["curl", "-fSL", "-o", str(dest), "--progress-bar"]
    if token:
        cmd.extend(["-H", f"Authorization: Bearer {token}"])
    cmd.append(url)

    result = subprocess.run(cmd, capture_output=True, text=True)
    if result.returncode != 0:
        dest.unlink(missing_ok=True)
        raise RuntimeError(f"Download failed: {result.stderr.strip()}")

File: src/hub.rs
//! HuggingFace Hub model download support.
//!
//! Downloads model files (config.json, tokenizer.json, safetensors) from
//! the HuggingFace Hub using the HTTP API. Supports caching to avoid
//! re-downloading.

use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

use log::info;
use thiserror::Error;

#[derive(Debug, Error)]
pub enum HubError {
    #[error("IO error: {0}")]
    Io(#[from] std::io::Error),
    #[error("HTTP error: {0}")]
    Http(String),
    #[error("model not found: {0}")]
    NotFound(String),
    #[error("huggingface-cli not available and HTTP download failed: {0}")]
    NoDownloadMethod(String),
    #[error("invalid model ID: {0}")]
    InvalidModelId(String),
}

/// Parsed model identifier (owner/name or just name).
#[derive(Debug, Clone)]
pub struct ModelId {
    pub repo_id: String,
    pub revision: String,
}

impl ModelId {
    pub fn parse(input: &str) -> Result<Self, HubError> {
        let input = input.trim();
        // Check if it's a local path.
        if Path::new(input).exists() {
            return Err(HubError::InvalidModelId(
                "input is a local path, not a model ID".into(),
            ));
        }

        let (repo, rev) = if input.contains('@') {
            let parts: Vec<&str> = input.splitn(2, '@').collect();
            (parts[0].to_string(), parts[1].to_string())
        } else {
            (input.to_string(), "main".to_string())
        };

        // Basic validation: should contain at most one slash.
        if repo.matches('/').count() > 1 {
            return Err(HubError::InvalidModelId(format!(
                "invalid repo ID: '{}'", repo
            )));
        }

        Ok(ModelId {
            repo_id: repo,
            revision: rev,
        })
    }
}

/// Resolve a model path — either a local directory or a HuggingFace model ID.
/// If it's a model ID, downloads to cache_dir and returns the cached path.
/// `hf_token` takes precedence over the HF_TOKEN / HUGGING_FACE_HUB_TOKEN env vars.
pub fn resolve_model_path(
    model_path: &str,
    cache_dir: Option<&Path>,
    hf_token: Option<&str>,
) -> Result<PathBuf, HubError> {
    let path = Path::new(model_path);

    // If it's already a local directory with config.json, use it directly.
    if path.is_dir() && path.join("config.json").exists() {
        return Ok(path.to_path_buf());
    }

    // Parse as a HuggingFace model ID.
    let model_id = ModelId::parse(model_path)?;

    // Determine cache directory.
    let cache = cache_dir
        .map(|p| p.to_path_buf())
        .unwrap_or_else(default_cache_dir);

    let model_cache = cache
        .join("models")
        .join(model_id.repo_id.replace('/', "--"));

    // Check if already cached.
    if model_cache.join("config.json").exists() {
        info!("Using cached model at {:?}", model_cache);
        return Ok(model_cache);
    }

    // Download the model.
    info!("Downloading model '{}' to {:?}", model_id.repo_id, model_cache);
    download_model(&model_id, &model_cache, hf_token)?;

    Ok(model_cache)
}

fn default_cache_dir() -> PathBuf {
    // Use HF_HOME or ~/.cache/nve
    if let Ok(hf_home) = std::env::var("HF_HOME") {
        PathBuf::from(hf_home).join("nve")
    } else if let Ok(xdg) = std::env::var("XDG_CACHE_HOME") {
        PathBuf::from(xdg).join("nve")
    } else if let Ok(home) = std::env::var("HOME") {
        PathBuf::from(home).join(".cache").join("nve")
    } else {
        PathBuf::from("/tmp/nve_cache")
    }
}

/// Download a model from HuggingFace Hub.
fn download_model(model_id: &ModelId, dest: &Path, hf_token: Option<&str>) -> Result<(), HubError> {
    // Try huggingface-cli first (handles auth, LFS, etc.).
    if try_hf_cli_download(model_id, dest)? {
        return Ok(());
    }

    // Fall back to direct HTTP download.
    download_via_http(model_id, dest, hf_token)
}

fn try_hf_cli_download(model_id: &ModelId, dest: &Path) -> Result<bool, HubError> {
    // Check if huggingface-cli is available.
    let check = Command::new("huggingface-cli").arg("--help").output();
    if check.is_err() {
        return Ok(false);
    }

    info!("Using huggingface-cli to download {}", model_id.repo_id);
    fs::create_dir_all(dest)?;

    let mut cmd = Command::new("huggingface-cli");
    cmd.arg("download")
        .arg(&model_id.repo_id)
        .arg("--local-dir")
        .arg(dest)
        .arg("--revision")
        .arg(&model_id.revision);

    // Only download essential files.
    cmd.arg("--include")
        .arg("config.json")
        .arg("--include")
        .arg("tokenizer.json")
        .arg("--include")
        .arg("tokenizer_config.json")
        .arg("--include")
        .arg("*.safetensors")
        .arg("--include")
        .arg("*.safetensors.index.json");

    let output = cmd.output()?;
    if output.status.success() {
        info!("Download complete via huggingface-cli");
        Ok(true)
    } else {
        let stderr = String::from_utf8_lossy(&output.stderr);
        info!("huggingface-cli failed: {}", stderr.trim());
        Ok(false)
    }
}

fn download_via_http(model_id: &ModelId, dest: &Path, hf_token: Option<&str>) -> Result<(), HubError> {
    fs::create_dir_all(dest)?;

    let base_url = format!(
        "https://huggingface.co/{}/resolve/{}",
        model_id.repo_id, model_id.revision
    );

    // Explicit token takes precedence over env vars.
    let auth_token = hf_token
        .map(|t| t.to_string())
        .or_else(|| std::env::var("HF_TOKEN").ok())
        .or_else(|| std::env::var("HUGGING_FACE_HUB_TOKEN").ok());

    // Required files.
    let required = vec!["config.json", "tokenizer.json"];
    // Optional files to try.
    let optional = vec![
        "tokenizer_config.json",
        "model.safetensors",
        "model.safetensors.index.json",
    ];

    for filename in &required {
        let url = format!("{}/{}", base_url, filename);
        let dest_path = dest.join(filename);
        download_file(&url, &dest_path, auth_token.as_deref()).map_err(|e| {
            HubError::Http(format!("failed to download {}: {}", filename, e))
        })?;
    }

    for filename in &optional {
        let url = format!("{}/{}", base_url, filename);
        let dest_path = dest.join(filename);
        let _ = download_file(&url, &dest_path, auth_token.as_deref());
    }

    // If we got an index file, download the shard files.
    let index_path = dest.join("model.safetensors.index.json");
    if index_path.exists() {
        let index_json = fs::read_to_string(&index_path)?;
        let index: serde_json::Value = serde_json::from_str(&index_json)
            .map_err(|e| HubError::Http(format!("invalid index JSON: {}", e)))?;

        if let Some(weight_map) = index.get("weight_map").and_then(|v| v.as_object()) {
            let shard_files: std::collections::HashSet<&str> = weight_map
                .values()
                .filter_map(|v| v.as_str())
                .collect();

            for shard in shard_files {
                let shard_dest = dest.join(shard);
                if !shard_dest.exists() {
                    let url = format!("{}/{}", base_url, shard);
                    info!("Downloading shard: {}", shard);
                    download_file(&url, &shard_dest, auth_token.as_deref()).map_err(|e| {
                        HubError::Http(format!("failed to download shard {}: {}", shard, e))
                    })?;
                }
            }
        }
    }

    Ok(())
}

fn download_file(url: &str, dest: &Path, auth_token: Option<&str>) -> Result<(), HubError> {
    info!("Downloading: {} -> {:?}", url, dest);

    // Use curl since it's universally available and handles redirects/LFS.
    let mut cmd = Command::new("curl");
    cmd.arg("-fSL") // fail on HTTP errors, show errors, follow redirects
        .arg("-o")
        .arg(dest)
        .arg("--progress-bar");

    if let Some(token) = auth_token {
        cmd.arg("-H").arg(format!("Authorization: Bearer {}", token));
    }

    cmd.arg(url);

    let output = cmd.output()?;
    if output.status.success() {
        Ok(())
    } else {
        // Clean up partial download.
        let _ = fs::remove_file(dest);
        let stderr = String::from_utf8_lossy(&output.stderr);
        Err(HubError::Http(format!("curl failed: {}", stderr.trim())))
    }
}

/// Check if a path looks like a HuggingFace model ID (org/model or just model).
pub fn is_model_id(s: &str) -> bool {
    !Path::new(s).exists() && !s.starts_with('/') && !s.starts_with('.')
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_model_id() {
        let id = ModelId::parse("meta-llama/Llama-3.2-1B").unwrap();
        assert_eq!(id.repo_id, "meta-llama/Llama-3.2-1B");
        assert_eq!(id.revision, "main");
    }

    #[test]
    fn test_parse_model_id_with_revision() {
        let id = ModelId::parse("meta-llama/Llama-3.2-1B@refs/pr/123").unwrap();
        assert_eq!(id.repo_id, "meta-llama/Llama-3.2-1B");
        assert_eq!(id.revision, "refs/pr/123");
    }
}
